consumer/consumererror/transporterrors.go

```diff
@@ -0,0 +1,91 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"fmt"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// TransportError contains either an HTTP or gRPC status code
+// resulting from a failed synchronous network request.
+// It allows retrieving a type of error code, converting
+// between the status codes supported by each transport if necessary.
+//
+// It should be created with NewHTTPStatus or NewGRPCStatus.
+type TransportError struct {
```

Comment:
Looks like we don't link the transport error with retriable errors in any way, right? Are the exporters expected to read codes and wrap the errors as retriable when needed? Why can't we define such mapping in this package?

Reply:
>Looks like we don't link the transport error with retriable errors in any way, right?

These would be linked with `errors.Join`, then upstream components can obtain the errors using `errors.As`. I expect these will always be related, but if you think it's possible for us to get unrelated errors in the chain we could find a way to link these.

>Are the exporters expected to read codes and wrap the errors as retriable when needed?

The expectation is for exporters to determine based on their protocol what's considered retryable. Our current assumptions are based on OTLP, but there may be other considerations for other protocols.

>Why can't we define such mapping in this package?

I think the mapping would be better suited for whatever public package the `internal/statusconversion` package becomes, but I would be open to having an `IsRetryable(*status.Status)` function or similar.

Reply:
I think default mapping is good as long as users can override it, which I believe they can just by wrapping a network error into a retryable or permanent one, right?

Reply:
Component authors have full control over which errors are considered retryable by combining a transport error and retryable error with `errors.Join` or just leaving out the retryable information to make it permanent. Users could only override this if an exporter that emitted the error allowed configuring which codes should be retried.